x-pack/plugins/security_solution/public/detections/pages/detection_engine/rules/all/actions.ts

```diff
@@ -56,28 +76,66 @@ export const executeRulesBulkAction = async ({
   onSuccess = defaultSuccessHandler,
   onError = defaultErrorHandler,
   onFinish,
-}: ExecuteRulesBulkActionArgs) => {
+}: RulesBulkActionArgs | ExportRulesBulkActionArgs) {
+  let response: Blob | BulkActionResponse | null = null;
   try {
     setLoadingRules?.({ ids: visibleRuleIds, action });
 
     if (action === BulkAction.export) {
-      const response = await performBulkAction({ ...search, action });
-      downloadBlob(response, `${i18n.EXPORT_FILENAME}.ndjson`);
-      onSuccess(toasts, action, await getExportedRulesCounts(response));
+      // on successToast for export handles separately outside of action execution method
+      response = await performBulkAction({ ...search, action });
     } else {
-      const response = await performBulkAction({ ...search, action, edit: payload?.edit });
+      response = await performBulkAction({ ...search, action, edit: payload?.edit });
       sendTelemetry(action, response);
       onSuccess(toasts, action, response.attributes.summary);
-
-      return response;
     }
   } catch (error) {
     onError(toasts, action, error);
   } finally {
     setLoadingRules?.({ ids: [], action: null });
     onFinish?.();
   }
-};
+  return response;
+}
+
+/**
+ * downloads exported rules, received from export action
+ * @param params.response - Blob results with exported rules
+ * @param params.toasts - {@link UseAppToasts} toasts service
+ * @param params.onSuccess - {@link OnActionSuccessCallback} optional toast to display when action successful
+ * @param params.onError - {@link OnActionErrorCallback} optional toast to display when action failed
+ */
+export async function downloadExportedRules({
+  response,
+  toasts,
+  onSuccess = defaultSuccessHandler,
+  onError = defaultErrorHandler,
+}: {
+  response: Blob;
+  toasts: UseAppToasts;
+  onSuccess?: OnActionSuccessCallback;
+  onError?: OnActionErrorCallback;
+}) {
+  try {
+    downloadBlob(response, `${i18n.EXPORT_FILENAME}.ndjson`);
+    onSuccess(toasts, BulkAction.export, await getExportedRulesCounts(response));
+  } catch (error) {
+    onError(toasts, BulkAction.export, error);
+  }
+}
+
+/**
+ * executes bulk export action and downloads exported rules
+ * @param params - {@link ExportRulesBulkActionArgs}
+ */
+export async function bulkExportRules(params: ExportRulesBulkActionArgs) {
+  const response = await executeRulesBulkAction(params);
+
+  // if response null, likely network error happened and export rules haven't been received
+  if (response) {
+    await downloadExportedRules({ response, toasts: params.toasts, onSuccess: params.onSuccess });
+  }
```

Comment:
I moved it out of `executeRulesBulkAction` because, apart from API call, it was doing download of exported rules file on action - `export`.

Now, when this download can be optional, I think it make sense to move it out. And `executeBulkAction` will only perform  API call and return response. 
When needed, this response can be processed accordingly to needs of the caller.

Reply:
Ok, that makes sense to me 👍